Adding worker nodes to the OCP 4 UPI cluster existing 24+ hours

Kuldip-Nanda opened the pull request:
1. If the cluster is older than a day, the bootstrap certificate has to be updated. This code takes care of this issue.
2. If the worker node count is decreased then the removed worker node is put in NotReady state. The loop where CSR is approved takes care of this.
3. kubeconfig is made available in bastion's ~/.kube directory in the beginning itself, otherwise the oc command does not work.

Commit:
Checking the cert. expiry time and cluster age to determine if it should be regenerated. Update the worker.ign file with the regenerated certificate and upload it to the HTTP server in use for the RH OCP.
Also removed oc checks and made a noclobber backup of worker.ign.original

yussufsh commented on playbooks/roles/ocp-install/tasks/checkcert.yaml:
```diff
@@ -0,0 +1,32 @@
+# update worker.ign file with new cert on HTTP Server if cluster age > 24 hrs(under24=0).
+- name: Update a certificate if required
+  shell: |
+    hostname=`hostname`
```
There should an env variable available called `HOSTNAME` can we use it directly?

Kuldip-Nanda replied:
sure. Let me take a look at this.

yussufsh commented on playbooks/roles/ocp-install/tasks/checkcert.yaml:
```diff
@@ -0,0 +1,32 @@
+- name: Update a certificate if required
+  shell: |
+    hostname=`hostname`
+    MCS=api-int.${hostname%"-bastion-0.ibm.com"}.ibm.com:22623
```
You should not hard-code the domain here to ibm.com. Also, I think it is hard way to get the sub-domain.
Instead I suggest you use variables `install_config.cluster_domain` eg: ibm.com and `install_config.cluster_id` eg: yus-f0a1 OR getting it form previous ignition file OR `oc config` may give you that api server URL?

Kuldip-Nanda replied:
Good catch. I just plainly forgot. Let me take a look at this.

yussufsh commented on playbooks/roles/ocp-install/tasks/checkcert.yaml:
```diff
@@ -0,0 +1,32 @@
+    tee api-int.base64 && sed --regexp-extended --in-place=.backup "s%base64,[^,]+%base64,$(cat api-int.base64)\"}]}}%" openstack-upi/worker.ign
+
+    # Copy Ignition Files to HTTP server
+    cp openstack-upi/worker.ign /var/www/html/ignition
```
This should be a force copy.

yussufsh commented on playbooks/roles/ocp-install/tasks/checkcert.yaml:
```diff
@@ -0,0 +1,32 @@
+    # get the cluster create time in seconds
+    tmstmp=`date -d$(oc get -o json clusterversion version |  jq -r '.metadata.creationTimestamp') '+%s'`
+    echo $(( $tmstmp  + 86400 > `date +%s` ? 1 : 0 ))
+  register: under24 
```
Should not the variable mean `over24hours`?

Kuldip-Nanda replied:
Well can be done. I was checking against Not Under 24hrs Condition.

yussufsh commented on playbooks/roles/ocp-install/tasks/main.yaml:
```diff
@@ -11,11 +11,13 @@
   shell: |
     oc get csr -ojson | jq -r '.items[] | select(.status == {} ) | .metadata.name' | xargs -r oc adm certificate approve
-  until: lookup('pipe','oc get nodes | grep -w worker | grep -w Ready | wc -l') == worker_count
+  until: lookup('pipe','oc get nodes | grep -w worker | grep -w Ready | grep  -v "NotReady" | wc -l') == worker_count
```
Any good reason for adding this? I am sure it means the same thing. However, I will go with what was working before.

Kuldip-Nanda replied:
So if one goes from 4 worker nodes to three worker nodes, the oc get node will put the 4th node in NotReady State. The command "grep -v"  excludes all the Nodes which are in NotReady State. This code should remain as is. I have tested going from 2 nodes to 6 nodes and back 2 nodes over a period of 3 days.

yussufsh replied:
The previous check is `grep -w Ready` so it won't count NotReady worker nodes.

Kuldip-Nanda replied:
Thanks. will do

bpradipt commented on playbooks/roles/ocp-install/tasks/checkcert.yaml:
```diff
@@ -0,0 +1,33 @@
+# update worker.ign file with new cert on HTTP Server if cluster age > 24 hrs(under24=0).
+- name: Update a certificate if required
+  shell: |
+    hostname=`hostname`
```
hostname can be removed now as it's not used ?

bpradipt commented:
@Kuldip-Nanda please squash all your commits into a single commit and re-push. Good link on squashing commits - https://www.educative.io/edpresso/how-to-squash-commits-in-git

yussufsh commented:
@Kuldip-Nanda does this code work for new workers added after 24 hours? I am worried because ocp-install role is run after nodes-config which checks for ssh connection to the worker nodes in the install play. Does the new worker nodes boot with older ignition file?

Kuldip-Nanda commented:
> @Kuldip-Nanda does this code work for new workers added after 24 hours? I am worried because ocp-install role is run after nodes-config which checks for ssh connection to the worker nodes in the install play. Does the new worker nodes boot with older ignition file?

Yes did testing on this and it worked on a cluster. I am doing more testing

Kuldip-Nanda commented:
> @Kuldip-Nanda please squash all your commits into a single commit and re-push. Good link on squashing commits - https://www.educative.io/edpresso/how-to-squash-commits-in-git

The changes are all squashed to single commit.

lsmcfadden commented:
@yussufsh @bpradipt did you review?  Any concerns or this one can be closed/merged?

Kuldip-Nanda commented:
1. I advanced the bastion server date by a day and ran the playbook from the bastion server to check if the ignition file is getting updated. I executed the terraform step -running playbook manually. 
```
[root@kn-034-bastion-0 openstack-upi]# cat /tmp/terraform_1793425539.sh
  #!/bin/sh
echo 'Running ocp install playbook...'
cd ocp4-playbooks && ansible-playbook -i inventory -e @install_vars.yaml playbooks/install.yaml -v
[root@kn-034-bastion-0 openstack-upi]# 
 ```

2. The play book ran smoothly and was able to accept a pending worker. as shown below:
```
TASK [ocp-install : include_tasks] ***********************************************************************************************************************************************************************************************************
included: /root/ocp4-playbooks/playbooks/roles/ocp-install/tasks/checkcert.yaml for kn-034-bastion-0

TASK [ocp-install : check if oc is installed] ************************************************************************************************************************************************************************************************
changed: [kn-034-bastion-0] => {"changed": true, "cmd": ["type", "-P", "oc"], "delta": "0:00:00.002539", "end": "2020-12-11 15:33:34.056956", "rc": 0, "start": "2020-12-11 15:33:34.054417", "stderr": "", "stderr_lines": [], "stdout": "/usr/local/bin/oc", "stdout_lines": ["/usr/local/bin/oc"]}

TASK [ocp-install : check the cluster timestamp] *********************************************************************************************************************************************************************************************
changed: [kn-034-bastion-0] => {"changed": true, "cmd": "# get the cluster create time in seconds\ntmstmp=`date -d$(oc get -o json clusterversion version |  jq -r '.metadata.creationTimestamp') '+%s'`\necho $(( $tmstmp  + 86400 < `date +%s` ? 1 : 0 ))\n", "delta": "0:00:04.826667", "end": "2020-12-11 15:33:39.247948", "rc": 0, "start": "2020-12-11 15:33:34.421281", "stderr": "", "stderr_lines": [], "stdout": "1", "stdout_lines": ["1"]}

TASK [ocp-install : Update a certificate if required] ****************************************************************************************************************************************************************************************
changed: [kn-034-bastion-0] => {"changed": true, "cmd": "MCS=api-int.kn-034.ibm.com:22623\n\n# update the tls certificate in the worker ignition file\necho \"q\"                                                        | openssl s_client -connect $MCS  -showcerts                      | awk '/-----BEGIN CERTIFICATE-----/,/-----END CERTIFICATE-----/' | base64 --wrap=0                                                 | tee api-int.base64 && sed --regexp-extended --in-place=.backup \"s%base64,[^,]+%base64,$(cat api-int.base64)\\\"}]}}%\" openstack-upi/worker.ign\n\n# Copy Ignition Files to HTTP server\ncp -f openstack-upi/worker.ign /var/www/html/ignition\n", "delta": "0:00:00.025278", "end": "2020-12-11 15:33:39.619705", "rc": 0, "start": "2020-12-11 15:33:39.594427", "stderr": "depth=0 CN = system:machine-config-server\nverify error:num=20:unable to get local issuer certificate\nverify return:1\ndepth=0 CN = system:machine-config-server\nverify error:num=21:unable to verify the first certificate\nverify return:1\nDONE", "stderr_lines": ["depth=0 CN = system:machine-config-server", "verify error:num=20:unable to get local issuer certificate", "verify return:1", "depth=0 CN = system:machine-config-server", "verify error:num=21:unable to verify the first certificate", "verify return:1", "DONE"], "stdout": "LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSURXVENDQWtHZ0F3SUJBZ0lJZEMrN0hpQTJYK0V3RFFZSktvWklodmNOQVFFTEJRQXdKakVTTUJBR0ExVUUKQ3hNSmIzQmxibk5vYVdaME1SQXdEZ1lEVlFRREV3ZHliMjkwTFdOaE1CNFhEVEl3TVRJeE1EQXlNRGN4T0ZvWApEVE13TVRJd09EQXlNRGN5TUZvd0p6RWxNQ01HQTFVRUF4TWNjM2x6ZEdWdE9tMWhZMmhwYm1VdFkyOXVabWxuCkxYTmxjblpsY2pDQ0FTSXdEUVlKS29aSWh2Y05BUUVCQlFBRGdnRVBBRENDQVFvQ2dnRUJBTVROSVpiUitKWTIKMjRha20zT2VsdmszODBOMmpnZmRtQzhmNmhoNUZzdnYxUTF5WTlMQ2owOG9aa2h4VElmUy9NSm1tUCtoaWg1SwpWeHVqWk1ZODNFZE9LTWlEKzBGU3ZEVUtPc01XU1pmV1BkaHZ1cWxTbjNYZmZGdGpndmt5VjdUV28yd0ZXUXJtCmw1ZHNGNjd5UHZnczRkVysrVXpzeGxkZ1NnYnNaWFpoZzhOeU1xMDZjcUU3WUJUR3JyZEMrcE45VExSVjB1aUYKdjlySnRlU3FyTm8xQU9RK0xtcHllcEFIZWpqWmJVM2dDUFhaOFRrNDRjZXRiTDZPWjZWdkgydUw0NTVRV01LeApuQktXYkIwNUV0dmg5WCtQKzNhRGhwQWlGM3VvWlJHdGFhQ2xYVml5WGxkU1dNYWIyT0V5M0dZYnhqZGlkQStpCkJua002TnZOWG5NQ0F3RUFBYU9CaVRDQmhqQVRCZ05WSFNVRUREQUtCZ2dyQmdFRkJRY0RBVEFNQmdOVkhSTUIKQWY4RUFqQUFNQjBHQTFVZERnUVdCQlMwNERGYnQwbW5BTDU2L3ByaTNza2FaZ2RmM3pBZkJnTlZIU01FR0RBVwpnQlJKWk1uY2pNK2ZXYUpqQkdTNGg4WjFTSjV5OGpBaEJnTlZIUkVFR2pBWWdoWmhjR2t0YVc1MExtdHVMVEF6Ck5DNXBZbTB1WTI5dE1BMEdDU3FHU0liM0RRRUJDd1VBQTRJQkFRQlROaDFhQkZjaFhXRU1mVFJ3NmNVNzk5WlcKTUs3cUdiSXNBRVZpWFJxSVlEVnJWbHphSURsQjBQRjdPNEpiMzFPclI1S2FMYVVkUm12eUE4Z2M3c0JCNngxQwpadkUrMlN5eDg2RW1OYlc5MDdDZEVkUmZWRU1iVlJFVzFFZVZIZjd5NHhPT2FDcjl2d0ZKVVJjK1VlZ0h4Y3JmCmZ1VFlFZ25MZktqNTE4Rm5yMCtVUERPVWowVGpQclFkTVh0L29mV3d6RnZsbzczeE40Q1dKT2d0bm5xUmd0aXgKWVJ2RVczbUFYQUp0WmdvRmt0azBaQXZjZ3dNNi9Gb0Q3NDJIcUFWNDdSTFJlWmE3VGk3SFc4VFROSWRoaTlYVwpweUw0eERsWkxxaTNpODVWRUJXcGF3YjBtWElWaXRGSXhyQ2l1dW40eG5uek1WeGF5d2VNalJCSkFEUlAKLS0tLS1FTkQgQ0VSVElGSUNBVEUtLS0tLQo=", "stdout_lines": ["LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSURXVENDQWtHZ0F3SUJBZ0lJZEMrN0hpQTJYK0V3RFFZSktvWklodmNOQVFFTEJRQXdKakVTTUJBR0ExVUUKQ3hNSmIzQmxibk5vYVdaME1SQXdEZ1lEVlFRREV3ZHliMjkwTFdOaE1CNFhEVEl3TVRJeE1EQXlNRGN4T0ZvWApEVE13TVRJd09EQXlNRGN5TUZvd0p6RWxNQ01HQTFVRUF4TWNjM2x6ZEdWdE9tMWhZMmhwYm1VdFkyOXVabWxuCkxYTmxjblpsY2pDQ0FTSXdEUVlKS29aSWh2Y05BUUVCQlFBRGdnRVBBRENDQVFvQ2dnRUJBTVROSVpiUitKWTIKMjRha20zT2VsdmszODBOMmpnZmRtQzhmNmhoNUZzdnYxUTF5WTlMQ2owOG9aa2h4VElmUy9NSm1tUCtoaWg1SwpWeHVqWk1ZODNFZE9LTWlEKzBGU3ZEVUtPc01XU1pmV1BkaHZ1cWxTbjNYZmZGdGpndmt5VjdUV28yd0ZXUXJtCmw1ZHNGNjd5UHZnczRkVysrVXpzeGxkZ1NnYnNaWFpoZzhOeU1xMDZjcUU3WUJUR3JyZEMrcE45VExSVjB1aUYKdjlySnRlU3FyTm8xQU9RK0xtcHllcEFIZWpqWmJVM2dDUFhaOFRrNDRjZXRiTDZPWjZWdkgydUw0NTVRV01LeApuQktXYkIwNUV0dmg5WCtQKzNhRGhwQWlGM3VvWlJHdGFhQ2xYVml5WGxkU1dNYWIyT0V5M0dZYnhqZGlkQStpCkJua002TnZOWG5NQ0F3RUFBYU9CaVRDQmhqQVRCZ05WSFNVRUREQUtCZ2dyQmdFRkJRY0RBVEFNQmdOVkhSTUIKQWY4RUFqQUFNQjBHQTFVZERnUVdCQlMwNERGYnQwbW5BTDU2L3ByaTNza2FaZ2RmM3pBZkJnTlZIU01FR0RBVwpnQlJKWk1uY2pNK2ZXYUpqQkdTNGg4WjFTSjV5OGpBaEJnTlZIUkVFR2pBWWdoWmhjR2t0YVc1MExtdHVMVEF6Ck5DNXBZbTB1WTI5dE1BMEdDU3FHU0liM0RRRUJDd1VBQTRJQkFRQlROaDFhQkZjaFhXRU1mVFJ3NmNVNzk5WlcKTUs3cUdiSXNBRVZpWFJxSVlEVnJWbHphSURsQjBQRjdPNEpiMzFPclI1S2FMYVVkUm12eUE4Z2M3c0JCNngxQwpadkUrMlN5eDg2RW1OYlc5MDdDZEVkUmZWRU1iVlJFVzFFZVZIZjd5NHhPT2FDcjl2d0ZKVVJjK1VlZ0h4Y3JmCmZ1VFlFZ25MZktqNTE4Rm5yMCtVUERPVWowVGpQclFkTVh0L29mV3d6RnZsbzczeE40Q1dKT2d0bm5xUmd0aXgKWVJ2RVczbUFYQUp0WmdvRmt0azBaQXZjZ3dNNi9Gb0Q3NDJIcUFWNDdSTFJlWmE3VGk3SFc4VFROSWRoaTlYVwpweUw0eERsWkxxaTNpODVWRUJXcGF3YjBtWElWaXRGSXhyQ2l1dW40eG5uek1WeGF5d2VNalJCSkFEUlAKLS0tLS1FTkQgQ0VSVElGSUNBVEUtLS0tLQo="]}

TASK [ocp-install : Approve Worker CSRs] *****************************************************************************************************************************************************************************************************
FAILED - RETRYING: Approve Worker CSRs (30 retries left).
FAILED - RETRYING: Approve Worker CSRs (29 retries left).
FAILED - RETRYING: Approve Worker CSRs (28 retries left).
FAILED - RETRYING: Approve Worker CSRs (27 retries left).
FAILED - RETRYING: Approve Worker CSRs (26 retries left).
FAILED - RETRYING: Approve Worker CSRs (25 retries left).
FAILED - RETRYING: Approve Worker CSRs (24 retries left).
changed: [kn-034-bastion-0] => {"attempts": 8, "changed": true, "cmd": "oc get csr -ojson | jq -r '.items[] | select(.status == {} ) | .metadata.name' | xargs -r oc adm certificate approve\n", "delta": "0:00:00.456449", "end": "2020-12-11 15:42:01.231772", "rc": 0, "start": "2020-12-11 15:42:00.775323", "stderr": "", "stderr_lines": [], "stdout": "certificatesigningrequest.certificates.k8s.io/csr-cx6gt approved", "stdout_lines": ["certificatesigningrequest.certificates.k8s.io/csr-cx6gt approved"]}
```


3. The file worker.ign.backup was original ignition file and  worker.ign  is the new file which was generated as shown below:
```
[root@kn-034-bastion-0 openstack-upi]# date
Fri Dec 11 16:16:15 EST 2020

[root@kn-034-bastion-0 openstack-upi]# cat worker.ign.backup
{"ignition":{"config":{"merge":[{"source":"https://api-int.kn-034.ibm.com:22623/config/worker"}]},"security":{"tls":{"certificateAuthorities":[{"source":"data:text/plain;charset=utf-8;base64,LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSURFRENDQWZpZ0F3SUJBZ0lJVVNuT1phSnR6bHd3RFFZSktvWklodmNOQVFFTEJRQXdKakVTTUJBR0ExVUUKQ3hNSmIzQmxibk5vYVdaME1SQXdEZ1lEVlFRREV3ZHliMjkwTFdOaE1CNFhEVEl3TVRJeE1EQXlNRGN4T0ZvWApEVE13TVRJd09EQXlNRGN4T0Zvd0pqRVNNQkFHQTFVRUN4TUpiM0JsYm5Ob2FXWjBNUkF3RGdZRFZRUURFd2R5CmIyOTBMV05oTUlJQklqQU5CZ2txaGtpRzl3MEJBUUVGQUFPQ0FROEFNSUlCQ2dLQ0FRRUF1RlFXbG5Ya0JpL3QKbWFxNUwvUDRCR24xZjcyT0tUZUVKTGo3WDVtdGcrQUFKVUh4K25Vc1J0Wk1aREJWemZoQlZQUysxN2dsMkJzZwpXQktNNGFVUEYvbWhVWDZ3b2JLd2MzdGhqMXh5dUtFSHQ2Yk40Q1lDVzNQdFlXY1VmdXZVNnc3cVNkM1hrVldTCkNpbDhiZWhXNGtCTUFWUCtVajczWjQ3QmZjR3AzQmRDZnJxV01FZzQwVUJYbUE1U0x5dHVhS2x5M1JjOVFpblAKbGxQeStmcjRsL2tCM3RsaEMrby9PNGYzWVhhZVE4UjVLNDUxaXZjcW1RcHEzS1BiK0QxRHZMQW5WVG94VkFNagpScStjWkZQYzJuR2hsbzJOMjVVVThELzFEWFgya3Z3empMK2dWNFlkS013RnRWNEQ4bFU0UDk4SlY0RDcxTi8yCnBwdEdUT21ySVFJREFRQUJvMEl3UURBT0JnTlZIUThCQWY4RUJBTUNBcVF3RHdZRFZSMFRBUUgvQkFVd0F3RUIKL3pBZEJnTlZIUTRFRmdRVVNXVEozSXpQbjFtaVl3Umt1SWZHZFVpZWN2SXdEUVlKS29aSWh2Y05BUUVMQlFBRApnZ0VCQUUvYjhNRGNjdnlQcjRJTnlkNnE5VmQvZDZWS2JqYlVYM0lDTytzK0ZKcGhqdm4zcFJVTHhBVE5Bd3BhCkxBdXdLdVpaKzlMQ0I3b1h4QUZmK091cnRyN0xyNkM0L3N1RE9oV2pPcnpkR3cwYk9OODMyYWxEUjU4M3BtVG8KcWVwcHhocTRhY0dSVGIwQnhYWXdlWEw1RFAwaTVNM1JIUDA4bVk2SWJPTUNQY2FVeWVteXZKKzEvbUgxcEMrUwpTbUh1VzA4RHdlUjdId2RSMlVnMWUyUXlaSGk3c1hEUm1rUzdrQnJUZ2JuUmtwRk0wUUdhY0NWeVhBdHR3Rm5LCitidDdNRkR5UFpaaXlPaUN0QjU4Ry9lV21OdTZMSEZKdmRwUmRZanRETDgxcHR6NllBNXNHRmRnOWxUSXN3OXgKazFzb0RqS21XdXJvK2VNYjRJSk9zMy9XcHFvPQotLS0tLUVORCBDRVJUSUZJQ0FURS0tLS0tCg=="}]}},"version":"3.1.0"}}[root@kn-034-bastion-0 openstack-upi]# cat worker.ign.backup | jq
{
  "ignition": {
    "config": {
      "merge": [
        {
          "source": "https://api-int.kn-034.ibm.com:22623/config/worker"
        }
      ]
    },
    "security": {
      "tls": {
        "certificateAuthorities": [
          {
            "source": "data:text/plain;charset=utf-8;base64,LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSURFRENDQWZpZ0F3SUJBZ0lJVVNuT1phSnR6bHd3RFFZSktvWklodmNOQVFFTEJRQXdKakVTTUJBR0ExVUUKQ3hNSmIzQmxibk5vYVdaME1SQXdEZ1lEVlFRREV3ZHliMjkwTFdOaE1CNFhEVEl3TVRJeE1EQXlNRGN4T0ZvWApEVE13TVRJd09EQXlNRGN4T0Zvd0pqRVNNQkFHQTFVRUN4TUpiM0JsYm5Ob2FXWjBNUkF3RGdZRFZRUURFd2R5CmIyOTBMV05oTUlJQklqQU5CZ2txaGtpRzl3MEJBUUVGQUFPQ0FROEFNSUlCQ2dLQ0FRRUF1RlFXbG5Ya0JpL3QKbWFxNUwvUDRCR24xZjcyT0tUZUVKTGo3WDVtdGcrQUFKVUh4K25Vc1J0Wk1aREJWemZoQlZQUysxN2dsMkJzZwpXQktNNGFVUEYvbWhVWDZ3b2JLd2MzdGhqMXh5dUtFSHQ2Yk40Q1lDVzNQdFlXY1VmdXZVNnc3cVNkM1hrVldTCkNpbDhiZWhXNGtCTUFWUCtVajczWjQ3QmZjR3AzQmRDZnJxV01FZzQwVUJYbUE1U0x5dHVhS2x5M1JjOVFpblAKbGxQeStmcjRsL2tCM3RsaEMrby9PNGYzWVhhZVE4UjVLNDUxaXZjcW1RcHEzS1BiK0QxRHZMQW5WVG94VkFNagpScStjWkZQYzJuR2hsbzJOMjVVVThELzFEWFgya3Z3empMK2dWNFlkS013RnRWNEQ4bFU0UDk4SlY0RDcxTi8yCnBwdEdUT21ySVFJREFRQUJvMEl3UURBT0JnTlZIUThCQWY4RUJBTUNBcVF3RHdZRFZSMFRBUUgvQkFVd0F3RUIKL3pBZEJnTlZIUTRFRmdRVVNXVEozSXpQbjFtaVl3Umt1SWZHZFVpZWN2SXdEUVlKS29aSWh2Y05BUUVMQlFBRApnZ0VCQUUvYjhNRGNjdnlQcjRJTnlkNnE5VmQvZDZWS2JqYlVYM0lDTytzK0ZKcGhqdm4zcFJVTHhBVE5Bd3BhCkxBdXdLdVpaKzlMQ0I3b1h4QUZmK091cnRyN0xyNkM0L3N1RE9oV2pPcnpkR3cwYk9OODMyYWxEUjU4M3BtVG8KcWVwcHhocTRhY0dSVGIwQnhYWXdlWEw1RFAwaTVNM1JIUDA4bVk2SWJPTUNQY2FVeWVteXZKKzEvbUgxcEMrUwpTbUh1VzA4RHdlUjdId2RSMlVnMWUyUXlaSGk3c1hEUm1rUzdrQnJUZ2JuUmtwRk0wUUdhY0NWeVhBdHR3Rm5LCitidDdNRkR5UFpaaXlPaUN0QjU4Ry9lV21OdTZMSEZKdmRwUmRZanRETDgxcHR6NllBNXNHRmRnOWxUSXN3OXgKazFzb0RqS21XdXJvK2VNYjRJSk9zMy9XcHFvPQotLS0tLUVORCBDRVJUSUZJQ0FURS0tLS0tCg=="
          }
        ]
      }
    },
    "version": "3.1.0"
  }
}
[root@kn-034-bastion-0 openstack-upi]# cat worker.ign | jq
{
  "ignition": {
    "config": {
      "merge": [
        {
          "source": "https://api-int.kn-034.ibm.com:22623/config/worker"
        }
      ]
    },
    "security": {
      "tls": {
        "certificateAuthorities": [
          {
            "source": "data:text/plain;charset=utf-8;base64,LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSURXVENDQWtHZ0F3SUJBZ0lJZEMrN0hpQTJYK0V3RFFZSktvWklodmNOQVFFTEJRQXdKakVTTUJBR0ExVUUKQ3hNSmIzQmxibk5vYVdaME1SQXdEZ1lEVlFRREV3ZHliMjkwTFdOaE1CNFhEVEl3TVRJeE1EQXlNRGN4T0ZvWApEVE13TVRJd09EQXlNRGN5TUZvd0p6RWxNQ01HQTFVRUF4TWNjM2x6ZEdWdE9tMWhZMmhwYm1VdFkyOXVabWxuCkxYTmxjblpsY2pDQ0FTSXdEUVlKS29aSWh2Y05BUUVCQlFBRGdnRVBBRENDQVFvQ2dnRUJBTVROSVpiUitKWTIKMjRha20zT2VsdmszODBOMmpnZmRtQzhmNmhoNUZzdnYxUTF5WTlMQ2owOG9aa2h4VElmUy9NSm1tUCtoaWg1SwpWeHVqWk1ZODNFZE9LTWlEKzBGU3ZEVUtPc01XU1pmV1BkaHZ1cWxTbjNYZmZGdGpndmt5VjdUV28yd0ZXUXJtCmw1ZHNGNjd5UHZnczRkVysrVXpzeGxkZ1NnYnNaWFpoZzhOeU1xMDZjcUU3WUJUR3JyZEMrcE45VExSVjB1aUYKdjlySnRlU3FyTm8xQU9RK0xtcHllcEFIZWpqWmJVM2dDUFhaOFRrNDRjZXRiTDZPWjZWdkgydUw0NTVRV01LeApuQktXYkIwNUV0dmg5WCtQKzNhRGhwQWlGM3VvWlJHdGFhQ2xYVml5WGxkU1dNYWIyT0V5M0dZYnhqZGlkQStpCkJua002TnZOWG5NQ0F3RUFBYU9CaVRDQmhqQVRCZ05WSFNVRUREQUtCZ2dyQmdFRkJRY0RBVEFNQmdOVkhSTUIKQWY4RUFqQUFNQjBHQTFVZERnUVdCQlMwNERGYnQwbW5BTDU2L3ByaTNza2FaZ2RmM3pBZkJnTlZIU01FR0RBVwpnQlJKWk1uY2pNK2ZXYUpqQkdTNGg4WjFTSjV5OGpBaEJnTlZIUkVFR2pBWWdoWmhjR2t0YVc1MExtdHVMVEF6Ck5DNXBZbTB1WTI5dE1BMEdDU3FHU0liM0RRRUJDd1VBQTRJQkFRQlROaDFhQkZjaFhXRU1mVFJ3NmNVNzk5WlcKTUs3cUdiSXNBRVZpWFJxSVlEVnJWbHphSURsQjBQRjdPNEpiMzFPclI1S2FMYVVkUm12eUE4Z2M3c0JCNngxQwpadkUrMlN5eDg2RW1OYlc5MDdDZEVkUmZWRU1iVlJFVzFFZVZIZjd5NHhPT2FDcjl2d0ZKVVJjK1VlZ0h4Y3JmCmZ1VFlFZ25MZktqNTE4Rm5yMCtVUERPVWowVGpQclFkTVh0L29mV3d6RnZsbzczeE40Q1dKT2d0bm5xUmd0aXgKWVJ2RVczbUFYQUp0WmdvRmt0azBaQXZjZ3dNNi9Gb0Q3NDJIcUFWNDdSTFJlWmE3VGk3SFc4VFROSWRoaTlYVwpweUw0eERsWkxxaTNpODVWRUJXcGF3YjBtWElWaXRGSXhyQ2l1dW40eG5uek1WeGF5d2VNalJCSkFEUlAKLS0tLS1FTkQgQ0VSVElGSUNBVEUtLS0tLQo="
          }
        ]
      }
    },
    "version": "3.1.0"
  }
}
```

Kuldip-Nanda commented:
Below is the output of the Terraform apply **after increasing worker nodes from 4 to 5** for cluster older than 24 hrs.
Please see

1. BOX-1 : line 8-19 where the new Certificate is generated, applied and csr approved
2. BOX-2: showing the complete output of the run
3. BOX-3: showing oc get nodes from the bastion server  showing 31 min old worker-3

>>>  **BOX-1 : line 8-19 where the new Certificate is generated, applied and csr approved** 
```
  1 module.install.null_resource.install: Still creating... [16m31s elapsed]
  2 module.install.null_resource.install (remote-exec): ok: [kn-034-bastion-0] => {"changed": false, "gid": 0, "group": "root", "mode": "0755", "owner": "root", "path": "/root/.kube", "secontext": "unconfined_u:object_r:admin_home_t:s0",     "size": 33, "state": "directory", "uid": 0}
  3 
  4 module.install.null_resource.install (remote-exec): TASK [ocp-install : Install kubeconfig] ****************************************
  5 module.install.null_resource.install (remote-exec): ok: [kn-034-bastion-0] => {"changed": false, "checksum": "818db8aa2f5df8c08262deae86f808e1e0d63467", "dest": "/root/.kube/config", "gid": 0, "group": "root", "mode": "0644", "owner":     "root", "path": "/root/.kube/config", "secontext": "system_u:object_r:admin_home_t:s0", "size": 24652, "state": "file", "uid": 0}
  6 
  7 module.install.null_resource.install (remote-exec): TASK [ocp-install : include_tasks] *********************************************
  8 module.install.null_resource.install (remote-exec): included: /root/ocp4-playbooks/playbooks/roles/ocp-install/tasks/checkcert.yaml for kn-034-bastion-0
  9 
 10 module.install.null_resource.install (remote-exec): TASK [ocp-install : check if oc is installed] **********************************
 11 module.install.null_resource.install (remote-exec): changed: [kn-034-bastion-0] => {"changed": true, "cmd": ["type", "-P", "oc"], "delta": "0:00:00.002380", "end": "2020-12-11 17:31:24.582113", "rc": 0, "start": "2020-12-11 17:31:24.5    79733", "stderr": "", "stderr_lines": [], "stdout": "/usr/local/bin/oc", "stdout_lines": ["/usr/local/bin/oc"]}
 12 
 13 module.install.null_resource.install (remote-exec): TASK [ocp-install : check the cluster timestamp] *******************************
 14 module.install.null_resource.install (remote-exec): changed: [kn-034-bastion-0] => {"changed": true, "cmd": "# get the cluster create time in seconds\ntmstmp=`date -d$(oc get -o json clusterversion version |  jq -r '.metadata.creation    Timestamp') '+%s'`\necho $(( $tmstmp  + 86400 < `date +%s` ? 1 : 0 ))\n", "delta": "0:00:03.763393", "end": "2020-12-11 17:31:28.693420", "rc": 0, "start": "2020-12-11 17:31:24.930027", "stderr": "", "stderr_lines": [], "stdout": "1",     "stdout_lines": ["1"]}
 15 
 16 module.install.null_resource.install (remote-exec): TASK [ocp-install : Update a certificate if required] **************************
 17 module.install.null_resource.install (remote-exec): changed: [kn-034-bastion-0] => {"changed": true, "cmd": "MCS=api-int.kn-034.ibm.com:22623\n\n# update the tls certificate in the worker ignition file\necho \"q\"                                                            | openssl s_client -connect $MCS  -showcerts                      | awk '/-----BEGIN CERTIFICATE-----/,/-----END CERTIFICATE-----/' | base64 --wrap=0                                                 |     tee api-int.base64 && sed --regexp-extended --in-place=.backup \"s%base64,[^,]+%base64,$(cat api-int.base64)\\\"}]}}%\" openstack-upi/worker.ign\n\n# Copy Ignition Files to HTTP server\ncp -f openstack-upi/worker.ign /var/www/html/ig    nition\n", "delta": "0:00:00.025314", "end": "2020-12-11 17:31:29.073012", "rc": 0, "start": "2020-12-11 17:31:29.047698", "stderr": "depth=0 CN = system:machine-config-server\nverify error:num=20:unable to get local issuer certificat    e\nverify return:1\ndepth=0 CN = system:machine-config-server\nverify error:num=21:unable to verify the first certificate\nverify return:1\nDONE", "stderr_lines": ["depth=0 CN = system:machine-config-server", "verify error:num=20:unab    le to get local issuer certificate", "verify return:1", "depth=0 CN = system:machine-config-server", "verify error:num=21:unable to verify the first certificate", "verify return:1", "DONE"], "stdout": "LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0t    LS0tCk1JSURXVENDQWtHZ0F3SUJBZ0lJZEMrN0hpQTJYK0V3RFFZSktvWklodmNOQVFFTEJRQXdKakVTTUJBR0ExVUUKQ3hNSmIzQmxibk5vYVdaME1SQXdEZ1lEVlFRREV3ZHliMjkwTFdOaE1CNFhEVEl3TVRJeE1EQXlNRGN4T0ZvWApEVE13TVRJd09EQXlNRGN5TUZvd0p6RWxNQ01HQTFVRUF4TWNjM2x6ZE    dWdE9tMWhZMmhwYm1VdFkyOXVabWxuCkxYTmxjblpsY2pDQ0FTSXdEUVlKS29aSWh2Y05BUUVCQlFBRGdnRVBBRENDQVFvQ2dnRUJBTVROSVpiUitKWTIKMjRha20zT2VsdmszODBOMmpnZmRtQzhmNmhoNUZzdnYxUTF5WTlMQ2owOG9aa2h4VElmUy9NSm1tUCtoaWg1SwpWeHVqWk1ZODNFZE9LTWlEKzBGU3ZE    VUtPc01XU1pmV1BkaHZ1cWxTbjNYZmZGdGpndmt5VjdUV28yd0ZXUXJtCmw1ZHNGNjd5UHZnczRkVysrVXpzeGxkZ1NnYnNaWFpoZzhOeU1xMDZjcUU3WUJUR3JyZEMrcE45VExSVjB1aUYKdjlySnRlU3FyTm8xQU9RK0xtcHllcEFIZWpqWmJVM2dDUFhaOFRrNDRjZXRiTDZPWjZWdkgydUw0NTVRV01LeApuQk    tXYkIwNUV0dmg5WCtQKzNhRGhwQWlGM3VvWlJHdGFhQ2xYVml5WGxkU1dNYWIyT0V5M0dZYnhqZGlkQStpCkJua002TnZOWG5NQ0F3RUFBYU9CaVRDQmhqQVRCZ05WSFNVRUREQUtCZ2dyQmdFRkJRY0RBVEFNQmdOVkhSTUIKQWY4RUFqQUFNQjBHQTFVZERnUVdCQlMwNERGYnQwbW5BTDU2L3ByaTNza2FaZ2Rm    M3pBZkJnTlZIU01FR0RBVwpnQlJKWk1uY2pNK2ZXYUpqQkdTNGg4WjFTSjV5OGpBaEJnTlZIUkVFR2pBWWdoWmhjR2t0YVc1MExtdHVMVEF6Ck5DNXBZbTB1WTI5dE1BMEdDU3FHU0liM0RRRUJDd1VBQTRJQkFRQlROaDFhQkZjaFhXRU1mVFJ3NmNVNzk5WlcKTUs3cUdiSXNBRVZpWFJxSVlEVnJWbHphSURsQj    BQRjdPNEpiMzFPclI1S2FMYVVkUm12eUE4Z2M3c0JCNngxQwpadkUrMlN5eDg2RW1OYlc5MDdDZEVkUmZWRU1iVlJFVzFFZVZIZjd5NHhPT2FDcjl2d0ZKVVJjK1VlZ0h4Y3JmCmZ1VFlFZ25MZktqNTE4Rm5yMCtVUERPVWowVGpQclFkTVh0L29mV3d6RnZsbzczeE40Q1dKT2d0bm5xUmd0aXgKWVJ2RVczbUFY    QUp0WmdvRmt0azBaQXZjZ3dNNi9Gb0Q3NDJIcUFWNDdSTFJlWmE3VGk3SFc4VFROSWRoaTlYVwpweUw0eERsWkxxaTNpODVWRUJXcGF3YjBtWElWaXRGSXhyQ2l1dW40eG5uek1WeGF5d2VNalJCSkFEUlAKLS0tLS1FTkQgQ0VSVElGSUNBVEUtLS0tLQo=", "stdout_lines": ["LS0tLS1CRUdJTiBDRVJUS    UZJQ0FURS0tLS0tCk1JSURXVENDQWtHZ0F3SUJBZ0lJZEMrN0hpQTJYK0V3RFFZSktvWklodmNOQVFFTEJRQXdKakVTTUJBR0ExVUUKQ3hNSmIzQmxibk5vYVdaME1SQXdEZ1lEVlFRREV3ZHliMjkwTFdOaE1CNFhEVEl3TVRJeE1EQXlNRGN4T0ZvWApEVE13TVRJd09EQXlNRGN5TUZvd0p6RWxNQ01HQTFVRUF    4TWNjM2x6ZEdWdE9tMWhZMmhwYm1VdFkyOXVabWxuCkxYTmxjblpsY2pDQ0FTSXdEUVlKS29aSWh2Y05BUUVCQlFBRGdnRVBBRENDQVFvQ2dnRUJBTVROSVpiUitKWTIKMjRha20zT2VsdmszODBOMmpnZmRtQzhmNmhoNUZzdnYxUTF5WTlMQ2owOG9aa2h4VElmUy9NSm1tUCtoaWg1SwpWeHVqWk1ZODNFZE9LT    WlEKzBGU3ZEVUtPc01XU1pmV1BkaHZ1cWxTbjNYZmZGdGpndmt5VjdUV28yd0ZXUXJtCmw1ZHNGNjd5UHZnczRkVysrVXpzeGxkZ1NnYnNaWFpoZzhOeU1xMDZjcUU3WUJUR3JyZEMrcE45VExSVjB1aUYKdjlySnRlU3FyTm8xQU9RK0xtcHllcEFIZWpqWmJVM2dDUFhaOFRrNDRjZXRiTDZPWjZWdkgydUw0NTV    RV01LeApuQktXYkIwNUV0dmg5WCtQKzNhRGhwQWlGM3VvWlJHdGFhQ2xYVml5WGxkU1dNYWIyT0V5M0dZYnhqZGlkQStpCkJua002TnZOWG5NQ0F3RUFBYU9CaVRDQmhqQVRCZ05WSFNVRUREQUtCZ2dyQmdFRkJRY0RBVEFNQmdOVkhSTUIKQWY4RUFqQUFNQjBHQTFVZERnUVdCQlMwNERGYnQwbW5BTDU2L3Bya    TNza2FaZ2RmM3pBZkJnTlZIU01FR0RBVwpnQlJKWk1uY2pNK2ZXYUpqQkdTNGg4WjFTSjV5OGpBaEJnTlZIUkVFR2pBWWdoWmhjR2t0YVc1MExtdHVMVEF6Ck5DNXBZbTB1WTI5dE1BMEdDU3FHU0liM0RRRUJDd1VBQTRJQkFRQlROaDFhQkZjaFhXRU1mVFJ3NmNVNzk5WlcKTUs3cUdiSXNBRVZpWFJxSVlEVnJ    WbHphSURsQjBQRjdPNEpiMzFPclI1S2FMYVVkUm12eUE4Z2M3c0JCNngxQwpadkUrMlN5eDg2RW1OYlc5MDdDZEVkUmZWRU1iVlJFVzFFZVZIZjd5NHhPT2FDcjl2d0ZKVVJjK1VlZ0h4Y3JmCmZ1VFlFZ25MZktqNTE4Rm5yMCtVUERPVWowVGpQclFkTVh0L29mV3d6RnZsbzczeE40Q1dKT2d0bm5xUmd0aXgKW    VJ2RVczbUFYQUp0WmdvRmt0azBaQXZjZ3dNNi9Gb0Q3NDJIcUFWNDdSTFJlWmE3VGk3SFc4VFROSWRoaTlYVwpweUw0eERsWkxxaTNpODVWRUJXcGF3YjBtWElWaXRGSXhyQ2l1dW40eG5uek1WeGF5d2VNalJCSkFEUlAKLS0tLS1FTkQgQ0VSVElGSUNBVEUtLS0tLQo="]}
 18 
 19 module.install.null_resource.install (remote-exec): TASK [ocp-install : Approve Worker CSRs] ***************************************
 20 module.install.null_resource.install (remote-exec): FAILED - RETRYING: Approve Worker CSRs (30 retries left).
 21 module.install.null_resource.install: Still creating... [16m41s elapsed]
 22 module.install.null_resource.install: Still creating... [16m51s elapsed]
 23 module.install.null_resource.install: Still creating... [17m1s elapsed]
 24 module.install.null_resource.install: Still creating... [17m11s elapsed]
 25 module.install.null_resource.install: Still creating... [17m21s elapsed]
 26 module.install.null_resource.install: Still creating... [17m31s elapsed]
 27 module.install.null_resource.install: Still creating... [17m41s elapsed]
 28 module.install.null_resource.install: Still creating... [17m51s elapsed]
 29 module.install.null_resource.install: Still creating... [18m1s elapsed]
 30 module.install.null_resource.install: Still creating... [18m11s elapsed]
 31 module.install.null_resource.install: Still creating... [18m21s elapsed]
 32 module.install.null_resource.install: Still creating... [18m31s elapsed]
 33 module.install.null_resource.install (remote-exec): FAILED - RETRYING: Approve Worker CSRs (29 retries left).
 34 module.install.null_resource.install: Still creating... [18m41s elapsed]
 35 module.install.null_resource.install: Still creating... [18m51s elapsed]
 36 module.install.null_resource.install: Still creating... [19m1s elapsed]

103 module.install.null_resource.install: Still creating... [29m11s elapsed]
104 module.install.null_resource.install: Still creating... [29m21s elapsed]
105 module.install.null_resource.install: Still creating... [29m31s elapsed]
106 module.install.null_resource.install: Still creating... [29m41s elapsed]
107 module.install.null_resource.install: Still creating... [29m51s elapsed]
108 module.install.null_resource.install (remote-exec): FAILED - RETRYING: Approve Worker CSRs (23 retries left).
109 module.install.null_resource.install: Still creating... [30m1s elapsed]
110 module.install.null_resource.install: Still creating... [30m11s elapsed]
111 module.install.null_resource.install: Still creating... [30m21s elapsed]
112 module.install.null_resource.install: Still creating... [30m31s elapsed]
113 module.install.null_resource.install: Still creating... [30m41s elapsed]
114 module.install.null_resource.install: Still creating... [30m51s elapsed]
115 module.install.null_resource.install: Still creating... [31m1s elapsed]
116 module.install.null_resource.install (remote-exec): FAILED - RETRYING: Approve Worker CSRs (22 retries left).
117 module.install.null_resource.install: Still creating... [31m11s elapsed]
118 module.install.null_resource.install: Still creating... [31m21s elapsed]
119 module.install.null_resource.install: Still creating... [31m31s elapsed]
120 module.install.null_resource.install: Still creating... [31m41s elapsed]
121 module.install.null_resource.install: Still creating... [31m51s elapsed]
122 module.install.null_resource.install: Still creating... [32m1s elapsed]
123 module.install.null_resource.install: Still creating... [32m11s elapsed]
124 module.install.null_resource.install: Still creating... [32m21s elapsed]
125 module.install.null_resource.install: Still creating... [32m31s elapsed]
126 module.install.null_resource.install: Still creating... [32m41s elapsed]
127 module.install.null_resource.install (remote-exec): FAILED - RETRYING: Approve Worker CSRs (21 retries left).
128 module.install.null_resource.install: Still creating... [32m51s elapsed]
129 module.install.null_resource.install: Still creating... [33m1s elapsed]
130 module.install.null_resource.install: Still creating... [33m11s elapsed]
131 module.install.null_resource.install: Still creating... [33m21s elapsed]
132 module.install.null_resource.install: Still creating... [33m31s elapsed]
133 module.install.null_resource.install: Still creating... [33m41s elapsed]
134 module.install.null_resource.install: Still creating... [33m51s elapsed]
135 module.install.null_resource.install: Still creating... [34m1s elapsed]
136 module.install.null_resource.install: Still creating... [34m11s elapsed]
137 module.install.null_resource.install: Still creating... [34m21s elapsed]
138 module.install.null_resource.install: Still creating... [34m31s elapsed]
139 module.install.null_resource.install: Still creating... [34m41s elapsed]
140 module.install.null_resource.install: Still creating... [34m51s elapsed]
141 module.install.null_resource.install: Still creating... [35m1s elapsed]
142 module.install.null_resource.install (remote-exec): Error from server (Timeout): the server was unable to return a response in the time allotted, but may still be processing the request (get nodes)
143 module.install.null_resource.install (remote-exec): FAILED - RETRYING: Approve Worker CSRs (20 retries left).
144 module.install.null_resource.install: Still creating... [35m11s elapsed]
145 module.install.null_resource.install: Still creating... [35m21s elapsed]
146 module.install.null_resource.install: Still creating... [35m31s elapsed]
147 module.install.null_resource.install: Still creating... [35m41s elapsed]
148 module.install.null_resource.install: Still creating... [35m51s elapsed]
149 module.install.null_resource.install: Still creating... [36m1s elapsed]
150 module.install.null_resource.install: Still creating... [36m11s elapsed]
151 module.install.null_resource.install (remote-exec): FAILED - RETRYING: Approve Worker CSRs (19 retries left).
152 module.install.null_resource.install: Still creating... [36m21s elapsed]
153 module.install.null_resource.install: Still creating... [36m31s elapsed]
154 module.install.null_resource.install: Still creating... [36m41s elapsed]
155 module.install.null_resource.install: Still creating... [36m51s elapsed]
156 module.install.null_resource.install: Still creating... [37m1s elapsed]
157 module.install.null_resource.install: Still creating... [37m11s elapsed]
158 module.install.null_resource.install (remote-exec): FAILED - RETRYING: Approve Worker CSRs (18 retries left).
159 module.install.null_resource.install: Still creating... [37m21s elapsed]
160 module.install.null_resource.install: Still creating... [37m31s elapsed]
161 module.install.null_resource.install: Still creating... [37m41s elapsed]
162 module.install.null_resource.install: Still creating... [37m51s elapsed]
163 module.install.null_resource.install: Still creating... [38m1s elapsed]
164 module.install.null_resource.install: Still creating... [38m11s elapsed]
165 module.install.null_resource.install: Still creating... [38m21s elapsed]
166 module.install.null_resource.install (remote-exec): FAILED - RETRYING: Approve Worker CSRs (17 retries left).
167 module.install.null_resource.install: Still creating... [38m31s elapsed]
168 module.install.null_resource.install: Still creating... [38m41s elapsed]
169 module.install.null_resource.install: Still creating... [38m51s elapsed]
170 module.install.null_resource.install: Still creating... [39m1s elapsed]
171 module.install.null_resource.install: Still creating... [39m11s elapsed]
172 module.install.null_resource.install: Still creating... [39m21s elapsed]
173 module.install.null_resource.install: Still creating... [39m31s elapsed]
174 module.install.null_resource.install (remote-exec): FAILED - RETRYING: Approve Worker CSRs (16 retries left).
175 module.install.null_resource.install: Still creating... [39m41s elapsed]
176 module.install.null_resource.install: Still creating... [39m51s elapsed]
177 module.install.null_resource.install: Still creating... [40m1s elapsed]
178 module.install.null_resource.install: Still creating... [40m11s elapsed]
179 module.install.null_resource.install: Still creating... [40m21s elapsed]
180 module.install.null_resource.install: Still creating... [40m31s elapsed]
181 module.install.null_resource.install (remote-exec): FAILED - RETRYING: Approve Worker CSRs (15 retries left).
182 module.install.null_resource.install: Still creating... [40m41s elapsed]
183 module.install.null_resource.install: Still creating... [40m51s elapsed]
184 module.install.null_resource.install: Still creating... [41m1s elapsed]
185 module.install.null_resource.install: Still creating... [41m11s elapsed]
186 module.install.null_resource.install: Still creating... [41m21s elapsed]
187 module.install.null_resource.install: Still creating... [41m31s elapsed]
188 module.install.null_resource.install: Still creating... [41m41s elapsed]
189 module.install.null_resource.install (remote-exec): FAILED - RETRYING: Approve Worker CSRs (14 retries left).
190 module.install.null_resource.install: Still creating... [41m51s elapsed]
191 module.install.null_resource.install: Still creating... [42m1s elapsed]
192 module.install.null_resource.install: Still creating... [42m11s elapsed]
193 module.install.null_resource.install: Still creating... [42m21s elapsed]
194 module.install.null_resource.install: Still creating... [42m31s elapsed]
195 module.install.null_resource.install: Still creating... [42m41s elapsed]
196 module.install.null_resource.install: Still creating... [42m51s elapsed]
197 module.install.null_resource.install (remote-exec): changed: [kn-034-bastion-0] => {"attempts": 18, "changed": true, "cmd": "oc get csr -ojson | jq -r '.items[] | select(.status == {} ) | .metadata.name' | xargs -r oc adm certificate     approve\n", "delta": "0:00:00.252247", "end": "2020-12-11 17:55:37.404641", "rc": 0, "start": "2020-12-11 17:55:37.152394", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}
198 
199 module.install.null_resource.install (remote-exec): TASK [ocp-install : Wait for install-complete] *********************************
200 module.install.null_resource.install: Still creating... [43m1s elapsed]
201 module.install.null_resource.install (remote-exec): changed: [kn-034-bastion-0] => {"changed": true, "cmd": "openshift-install wait-for install-complete --log-level info", "delta": "0:00:08.277413", "end": "2020-12-11 17:55:46.257790"    , "rc": 0, "start": "2020-12-11 17:55:37.980377", "stderr": "level=info msg=\"Waiting up to 40m0s for the cluster at https://api.kn-034.ibm.com:6443 to initialize...\"\nlevel=info msg=\"Waiting up to 10m0s for the openshift-console ro    ute to be created...\"\nlevel=info msg=\"Install complete!\"\nlevel=info msg=\"To access the cluster as the system:admin user when using 'oc', run 'export KUBECONFIG=/root/openstack-upi/auth/kubeconfig'\"\nlevel=info msg=\"Access the     OpenShift web-console here: https://console-openshift-console.apps.kn-034.ibm.com\"\nlevel=info msg=\"Login to the console with user: \\\"kubeadmin\\\", and password: \\\"nouzS-Rxu5o-uZKaR-fNZq8\\\"\"\nlevel=info msg=\"Time elapsed: 8    s\"", "stderr_lines": ["level=info msg=\"Waiting up to 40m0s for the cluster at https://api.kn-034.ibm.com:6443 to initialize...\"", "level=info msg=\"Waiting up to 10m0s for the openshift-console route to be created...\"", "level=inf    o msg=\"Install complete!\"", "level=info msg=\"To access the cluster as the system:admin user when using 'oc', run 'export KUBECONFIG=/root/openstack-upi/auth/kubeconfig'\"", "level=info msg=\"Access the OpenShift web-console here: h    ttps://console-openshift-console.apps.kn-034.ibm.com\"", "level=info msg=\"Login to the console with user: \\\"kubeadmin\\\", and password: \\\"nouzS-Rxu5o-uZKaR-fNZq8\\\"\"", "level=info msg=\"Time elapsed: 8s\""], "stdout": "", "std    out_lines": []}
202 
203 module.install.null_resource.install (remote-exec): TASK [ocp-install : Configure nfs-provisioner] *********************************
204 module.install.null_resource.install: Still creating... [43m11s elapsed]
205 module.install.null_resource.install (remote-exec): changed: [kn-034-bastion-0] => {"changed": true, "cmd": "helpernodecheck nfs-setup", "delta": "0:00:04.207349", "end": "2020-12-11 17:55:50.808783", "rc": 0, "start": "2020-12-11 17:    55:46.601434", "stderr": "", "stderr_lines": [], "stdout": "ERROR: Looks like you've already deployed the nfs-provisioner", "stdout_lines": ["ERROR: Looks like you've already deployed the nfs-provisioner"]}
206 
207 module.install.null_resource.install (remote-exec): TASK [ocp-install : Wait for image-registry operator] **************************
208 module.install.null_resource.install (remote-exec): changed: [kn-034-bastion-0] => {"attempts": 1, "changed": true, "cmd": "oc get configs.imageregistry.operator.openshift.io/cluster --no-headers | wc -l", "delta": "0:00:00.314770", "    end": "2020-12-11 17:55:51.466160", "rc": 0, "start": "2020-12-11 17:55:51.151390", "stderr": "", "stderr_lines": [], "stdout": "1", "stdout_lines": ["1"]}
209 

```

>>> BOX-2: showing the complete output of the run
```
Apply complete! Resources: 2 added, 0 changed, 0 destroyed.

Outputs:

bastion_private_ip = 192.168.25.96
bastion_public_ip = 169.48.22.218
bastion_ssh_command = ssh -i data/id_rsa root@169.48.22.218
bootstrap_ip = 192.168.25.81
cluster_authentication_details = Cluster authentication details are available in 169.48.22.218 under ~/openstack-upi/auth
cluster_id = kn-034
etc_hosts_entries = 
169.48.22.218 api.kn-034.ibm.com console-openshift-console.apps.kn-034.ibm.com integrated-oauth-server-openshift-authentication.apps.kn-034.ibm.com oauth-openshift.apps.kn-034.ibm.com prometheus-k8s-openshift-monitoring.apps.kn-034.ibm.com grafana-openshift-monitoring.apps.kn-034.ibm.com example.apps.kn-034.ibm.com

install_status = COMPLETED
master_ips = [
  "192.168.25.113",
  "192.168.25.203",
  "192.168.25.156",
]
oc_server_url = https://api.kn-034.ibm.com:6443
storageclass_name = nfs-storage-provisioner
web_console_url = https://console-openshift-console.apps.kn-034.ibm.com
worker_ips = [
  "192.168.25.197",
  "192.168.25.147",
  "192.168.25.155",
  "192.168.25.161",
  "192.168.25.158",
]
```

>>> BOX-3: showing oc get nodes from the bastion server showing 31 min old worker-3
```
kuldips-mbp:ocp4-upi-powervs kuldip.nanda$ ssh -i data/id_rsa root@169.48.22.218 oc get nodes
Warning: untrusted X11 forwarding setup failed: xauth key data not generated
NAME       STATUS   ROLES    AGE     VERSION
master-0   Ready    master   21h     v1.19.0+43983cd
master-1   Ready    master   21h     v1.19.0+43983cd
master-2   Ready    master   21h     v1.19.0+43983cd
worker-0   Ready    worker   20h     v1.19.0+43983cd
worker-1   Ready    worker   20h     v1.19.0+43983cd
worker-2   Ready    worker   7h11m   v1.19.0+43983cd
worker-3   Ready    worker   168m    v1.19.0+43983cd
worker-4   Ready    worker   31m     v1.19.0+43983cd
```

Kuldip-Nanda commented:
Please see the output below after adding a worker in 5 days old cluster:
 
```
Outputs:

bastion_private_ip = 192.168.25.22
bastion_public_ip = 130.198.3.45
bastion_ssh_command = ssh -i data/id_rsa root@130.198.3.45
bootstrap_ip = 192.168.25.85
cluster_authentication_details = Cluster authentication details are available in 130.198.3.45 under ~/openstack-upi/auth
cluster_id = kn-601
etc_hosts_entries = 
130.198.3.45 api.kn-601.ibm.com console-openshift-console.apps.kn-601.ibm.com integrated-oauth-server-openshift-authentication.apps.kn-601.ibm.com oauth-openshift.apps.kn-601.ibm.com prometheus-k8s-openshift-monitoring.apps.kn-601.ibm.com grafana-openshift-monitoring.apps.kn-601.ibm.com example.apps.kn-601.ibm.com

install_status = COMPLETED
master_ips = [
  "192.168.25.165",
  "192.168.25.105",
  "192.168.25.109",
]
oc_server_url = https://api.kn-601.ibm.com:6443
storageclass_name = nfs-storage-provisioner
web_console_url = https://console-openshift-console.apps.kn-601.ibm.com
worker_ips = [
  "192.168.25.176",
  "192.168.25.192",
  "192.168.25.206",
  "192.168.25.101",
]
kuldips-mbp:ocp4-upi-powervs.syd kuldip.nanda$ ssh -i data/id_rsa root@130.198.3.45 oc get nodes
Warning: untrusted X11 forwarding setup failed: xauth key data not generated
NAME       STATUS   ROLES    AGE     VERSION
master-0   Ready    master   5d15h   v1.19.0+43983cd
master-1   Ready    master   5d15h   v1.19.0+43983cd
master-2   Ready    master   5d15h   v1.19.0+43983cd
worker-0   Ready    worker   5d15h   v1.19.0+43983cd
worker-1   Ready    worker   5d15h   v1.19.0+43983cd
worker-2   Ready    worker   5d15h   v1.19.0+43983cd
worker-3   Ready    worker   2m36s   v1.19.0+43983cd
kuldips-mbp:ocp4-upi-powervs.syd kuldip.nanda$ 
```

yussufsh commented:
I am yet to test it on a cluster running for more than a day. But I had asked @Prajyot-Parab to try your code where he was getting below error.
```
FAILED - RETRYING: Approve Worker CSRs (24 retries left).
Unable to connect to the server: x509: certificate has expired or is not yet valid: current time 2020-12-17T03:25:36-05:00 is after 2020-12-16T15:06:37Z
```

yussufsh commented:
Also, I see you are comparing the cluster time with current time. The ignition is always updated on every run and we could not compare the difference between the original and new worker.ign files. Let's investigate on the possibility to get the certificate validity instead and compare it with current time.

Another concern to merge this PR is that it will failed with a cluster where bootstrap node deleted. This will break other projects using this playbook and in turn the TF code.

I have a PR created for helpernode to support running without bootstrap node. Once that is upstream we can change the TF code to handle this usecase. Then this code will then make sense IMO.

Kuldip-Nanda commented:
Well as we had discussed and also I left a note for in the slack also, I tried to get the dates on the newly created Cluster from the openssl cert.
```
[root@kn-340-bastion-0 ~]# openssl  s_client -connect api-int.kn-340.ibm.com:22623 | openssl x509 -noout -dates
depth=0 CN = system:machine-config-server
verify error:num=20:unable to get local issuer certificate
verify return:1
depth=0 CN = system:machine-config-server
verify error:num=21:unable to verify the first certificate
verify return:1
notBefore=Dec 16 16:40:29 2020 GMT
notAfter=Dec 14 16:40:30 2030 GMT
read:errno=0
```
I am not sure if this is correct as I suspect the certificate has not been signed with start and end dates correctly. notAfter should be Dec 17 ... as notBefore is Dec 16....
Also Error unable to get local issuer certificate means, that the openssl does not know  root CA cert. I tried with -CApath as /etc/ssl/certs. It was not able to find the Trusted Certificate.
Can someone suggest on this.

Kuldip-Nanda commented:
1. For your concern about [output as seen by @Prajyot-Parab](https://github.com/ocp-power-automation/ocp4-playbooks/pull/92#issuecomment-747425420)  , please see my [output](https://github.com/ocp-power-automation/ocp4-playbooks/pull/92#issuecomment-742874831). So this is expected and will eventually resolve when the CSR presents on the host node. My test resolved in last few tries.
Also I did talk with Sridhar. They are using my patch and that they have successfully added a node multiple number of time.
2. Right now we do require bootstrap node. I did talk about it in the Scrum. Good to know this is being taken care of.
3. I have completed prototyping based on certificate validity. The playbook will recreate  and add a new certificate in the worker ignition file  if it determines the certificate has expired. I will be updating the PR after testing it.

yussufsh reviewed (dismissed):
/lgtm
For now changes are good to add new worker nodes. Going forward we can work on changes to the playbook to check the certificate validity against current time. Instead of comparing it with cluster up time.

bpradipt commented:
@Kuldip-Nanda this PR needs to be targeted for main/master branch.  Didn't realise it earlier.

bpradipt commented on playbooks/roles/ocp-install/tasks/checkcert.yaml:
```diff
@@ -0,0 +1,31 @@
+  register: over24
+  when: oc_installed.rc == 0
+
+# update worker.ign file with new cert on HTTP Server if cluster age > 24 hrs(under24=0).
```
We can add a link to this as comment - https://access.redhat.com/solutions/4799921
Also the var is over24 so the comment above `under24=0` should change to `over24==1`

Kuldip-Nanda replied:
Hi Pradipta,
I have made changes to look at the certificate validity dates to see if we have valid or expired certificates.
First run it looks good. Please review.

Kuldip-Nanda replied:
Definitely, I will add a link in the comment . 
Also the first run looks  great and it works. I will be testing these code changes with
1. add, remove add a node to a cluster when cluster is less than 24 hrs. old.
2. perfume the same task for this cluster when it is more than 24 hrs old.

The new changes extract the certificate from worker ignition file and get the validity dates of the certificate. Based on this it makes the determination if a new certificate is to be generated.

yussufsh commented:
> 
> 
> @Kuldip-Nanda this PR needs to be targeted for main/master branch. Didn't realise it earlier.

Good catch.. even I missed it.

Kuldip-Nanda commented:
Please see the latest code changes. These changes look at the certificate validity to check if the new certificate is to be inserted in the worker.ign.  On other note, I developed it on release-4.6   to test and validate. I can merge these changes in main branch after completing this iteration.

bpradipt commented on playbooks/roles/ocp-install/tasks/checkcert.yaml:
```diff
@@ -0,0 +1,44 @@
+
+    notBefore=$(echo "${dates}" | sed -n -e 's/^notBefore=//p' | date -d - +%s)
+    notAfter=$(echo "${dates}" | sed -n -e 's/^notAfter=//p' | date -d - +%s)
+    present=`date +%s`
```
This is good. @Kuldip-Nanda can you also check if you can use `checkend` option? That will do away with the sed logic
eg. `if openssl x509 -checkend 3600 -noout -in <certificate>` checks if certificate is good for 1 hr. Checking if certificate is valid for next 1 hr should be more than enough I think.

Also please raise the PR against the master/main branch.

Kuldip-Nanda replied:
I know we discussed this earlier but was not sure how much time I should add.  I did try this though after adding some time. On regeneration I get the old certificate.  I do not get the new certificate as I still have 1 hrs left when I try to regenerate the certificate. Looks like OpenSSL  generates only when the old one has expired. I will do some more testing on this.

yussufsh commented on playbooks/roles/ocp-install/tasks/checkcert.yaml:
```diff
@@ -0,0 +1,44 @@
+# tasks file for ocp4 install cert
+
+- name: check if oc is installed
+  command: type -P oc
```
We do not need oc now.

yussufsh commented on playbooks/roles/ocp-install/tasks/checkcert.yaml:
```diff
@@ -0,0 +1,44 @@
+       echo 0
+    fi
+  register: cert_valid
+  when: oc_installed.rc == 0
```
Remove when condition as oc is not required.

yussufsh reviewed (changes requested):
Overall changes look good. But I prefer the changes to be done in ocp-config role instead of ocp-install as we do not have dependency on oc. Also it may help to update all ignition files in `/var/www/html`. That way I think cluster can be installed with more than 24 hrs gap.

yussufsh commented on playbooks/roles/ocp-install/tasks/checkcert.yaml:
```diff
@@ -0,0 +1,44 @@
+    openssl s_client -connect $MCS  -showcerts                      | \
+    awk '/-----BEGIN CERTIFICATE-----/,/-----END CERTIFICATE-----/' | \
+    base64 --wrap=0                                                 | \
+    tee api-int.base64 && sed --regexp-extended --in-place=.backup "s%base64,[^,]+%base64,$(cat api-int.base64)\"}]}}%" openstack-upi/worker.ign
```
Can we copy the original ign file into worker.ign.orig? asking because multiple runs to add workers will keep on creating backup file over-writing the original contents.

Kuldip-Nanda replied:
fair enough. I will make an original copy for safe keeping. and each run will also create a backup file for the last run.

bpradipt commented:
> Please see the latest code changes. These changes look at the certificate validity to check if the new certificate is to be inserted in the worker.ign. On other note, I developed it on release-4.6 to test and validate. I can merge these changes in main branch after completing this iteration.

yeah, once all the changes are completed and tested, you can rebase it for master/main

Kuldip-Nanda commented:
I have added an additional check to see if the cluster is +24 hrs old. This will force it to regenerate the certificate, as I see the original certificate has validity of 10 years. Please sere the results below:
```
kuldips-mbp:temp kuldip.nanda$ openssl x509 -noout -in data.pem -dates
notBefore=Jan 13 19:35:57 2021 GMT
notAfter=Jan 11 19:35:57 2031 GMT
```

Below are the results on a cluster more than 24 hours old with 5 nodes. I added one more node:
```
Apply complete! Resources: 3 added, 0 changed, 2 destroyed.

Outputs:

bastion_private_ip = 192.168.25.56
bastion_public_ip = 130.198.121.43
bastion_ssh_command = ssh -i data/id_rsa root@130.198.121.43
bootstrap_ip = 192.168.25.248
cluster_authentication_details = Cluster authentication details are available in 130.198.121.43 under ~/openstack-upi/auth
cluster_id = kn-124
etc_hosts_entries = 
130.198.121.43 api.kn-124.ibm.com console-openshift-console.apps.kn-124.ibm.com integrated-oauth-server-openshift-authentication.apps.kn-124.ibm.com oauth-openshift.apps.kn-124.ibm.com prometheus-k8s-openshift-monitoring.apps.kn-124.ibm.com grafana-openshift-monitoring.apps.kn-124.ibm.com example.apps.kn-124.ibm.com

install_status = COMPLETED
master_ips = [
  "192.168.25.65",
  "192.168.25.31",
  "192.168.25.101",
]
oc_server_url = https://api.kn-124.ibm.com:6443
storageclass_name = nfs-storage-provisioner
web_console_url = https://console-openshift-console.apps.kn-124.ibm.com
worker_ips = [
  "192.168.25.116",
  "192.168.25.16",
  "192.168.25.169",
  "192.168.25.27",
  "192.168.25.15",
  "192.168.25.100",
]
```

yussufsh commented:
> I have added an additional check to see if the cluster is +24 hrs old. This will force it to regenerate the certificate, as I see the original certificate has validity of 10 years. Please sere the results below:
> 
> ```
> kuldips-mbp:temp kuldip.nanda$ openssl x509 -noout -in data.pem -dates
> notBefore=Jan 13 19:35:57 2021 GMT
> notAfter=Jan 11 19:35:57 2031 GMT
> ```
> 

If validity is 10 years then why do we need this change? I went to https://access.redhat.com/solutions/4799921 to check the comments where even they are mentioning that we do not need to update the worker ignition file.

I have tested the addition of worker nodes on a cluster with more than 6 days old. With just TF side changes I was able to add new workers to the existing cluster. Still wondering if this PR is needed?

I will do few more tests and raise PR for TF itself for adding and deleting the node.

Kuldip-Nanda commented:
> > I have added an additional check to see if the cluster is +24 hrs old. This will force it to regenerate the certificate, as I see the original certificate has validity of 10 years. Please sere the results below:
> > ```
> > kuldips-mbp:temp kuldip.nanda$ openssl x509 -noout -in data.pem -dates
> > notBefore=Jan 13 19:35:57 2021 GMT
> > notAfter=Jan 11 19:35:57 2031 GMT
> > ```
> 
> If validity is 10 years then why do we need this change? I went to https://access.redhat.com/solutions/4799921 to check the comments where even they are mentioning that we do not need to update the worker ignition file.
> 
> I have tested the addition of worker nodes on a cluster with more than 6 days old. With just TF side changes I was able to add new workers to the existing cluster. Still wondering if this PR is needed?
> 
> I will do few more tests and raise PR for TF itself for adding and deleting the node.


Yes, this is surprising as when  we initially started, we were not able to add a node without updating the ignition file [slack ref:  [Adding a node to a cluster:](https://powercoop.slack.com/archives/CLB0VTQF5/p1602187877066300)]. Some improvements must have been made in the CoreOS.

However my thinking on the other check - `Check the certificate expiry time` is that it is a valid check and should be introduced. 

I already have a Terraform PR opened but is pending the approval of this PR. Please see
  [https://github.com/ocp-power-automation/ocp4-upi-powervs/pull/156](https://github.com/ocp-power-automation/ocp4-upi-powervs/pull/156)
  
This contains the  accompanying terraform changes.

Kuldip-Nanda commented:
Based on the feedback from RH folks as below we do not require any changes in the ocp4-playbook. This can be closed as not required:
~~~
the ignitions for the workers and masters contain a ca that expires in 10 years i believe.  you can use these ignition files as they are to add worker nodes with no issue.  There is a CA cert in the master and worker ignitions.  only the bootstrap ignition file has a short expiration.
the ignitions are generated by the installer and are immutable(i.e. you cant customize cert expirations).  after install, the cluster manages and rotates certificates
~~~
